### PR TITLE
Vrouter static configuration

### PR DESCRIPTION
**Pull request by psykulsk**
In the Openshift side-by-side deployment scenario, Contrail control components are deployed outside of the Openshift cluster, so the behaviour of the contrail-operator has to be changed.

Current state: Vrouter Custom Resource and its controller in contrail-operator, support only dynamic configuration. Controller finds out the IPs of the control, collector etc. nodes and generates a config for the vrouter with them. 

Required state: User has an option to provide static configuration of the Vrouter (IP’s of the contrail control, collector, etc.) via the definition of the Vrouter Custom Resource

### CODE REVIEW

**Comment by psykulsk** [comment]
recheck

**Comment by psykulsk** [comment]
recheck

**Comment by Dysproz** [comment]
recheck

**Comment by psykulsk** [comment]
recheck ha

**Review by Dysproz — CHANGES_REQUESTED** [review]

File: pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go [hunk 1]
```diff
@@ -173,3 +287,191 @@ fabric_snat_hash_table_size = 4096
 [SESSION]
 slo_destination = collector
 sample_destination = collector`
+
+var vrouterConfigStaticConfigurationConfigControlDefaultPorts = `[CONTROL-NODE]
+servers=1.1.1.1:5269 2.2.2.2:5269 3.3.3.3:5269
```
[the remark this answers is not in the document]
Comment (psykulsk) [reply pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go:292]: Yep, good idea. I will do that.

File: pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go [hunk 2]
```diff
@@ -127,6 +129,118 @@ func TestVrouterCustomEnvVariablesConfigMap(t *testing.T) {
...
+				}
+				if err := cl.Get(context.TODO(), types.NamespacedName{Name: "vrouter1-vrouter-configmap", Namespace: "default"}, &environment.vrouterConfigMap); err != nil {
+					t.Fatalf("get configmap: (%v)", err)
+				}
+				if environment.vrouterConfigMap.Data["vrouter.1.1.8.1"] != vrouterConfigStaticConfigurationControlDefaultPorts {
+					configDiff := diff.Diff(environment.vrouterConfigMap.Data["vrouter.1.1.8.1"], vrouterConfigStaticConfigurationControlDefaultPorts)
+					t.Fatalf("get vrouter config: \n%v\n", configDiff)
+				}
+			})
+		})
```
[the remark this answers is not in the document]
Comment (psykulsk) [reply pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go:240]: But there are cases for a fully dynamic config, a fully static config and from mixed configs, aren' there?

File: pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go [hunk 3]
```diff
@@ -127,6 +129,118 @@ func TestVrouterCustomEnvVariablesConfigMap(t *testing.T) {
...
+				ControlNodesConfiguration: &v1alpha1.ControlClusterConfiguration{
+					ControlServerIPList: []string{"1.1.1.1", "2.2.2.2", "3.3.3.3"},
+				},
+			}
+
+			t.Run("then vrouter configmap has data with config and control ip's specified in the StaticConfiguration", func(t *testing.T) {
+				if err := environment.vrouterResource.InstanceConfiguration(reconcile.Request{NamespacedName: types.NamespacedName{Name: "vrouter1", Namespace: "default"}},
+					&environment.vrouterPodList, cl); err != nil {
+					t.Fatalf("get configmap: (%v)", err)
+				}
```
[the remark this answers is not in the document]
Comment (psykulsk) [reply pkg/apis/contrail/v1alpha1/tests/config_vrouter_test.go:153]: I think I prefer testing this way as it tests the exposed method's output, not the implementation inside the non exposed methods.

**Review by Dysproz — APPROVED** [review]

**Review by kaweue — APPROVED** [review]

**Pull request merged**